### PR TITLE
Add sklearn linear model example

### PR DESCRIPTION
**Pull request by ZiyueXu77**

### CODE REVIEW

**Comment by ZiyueXu77** [comment]
> Linear regression example should be simple. The example looks complicated. Can we make it simpler ?

I guess I can convert the the model to follow the "weight" format and reduce the aggregator/shareable_generator usage, essentially following the numpy example.

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/higgs_sklearn_linear_executor.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    data_size = end - start
+    data = pd.read_csv(data_path, header=None, skiprows=start, nrows=data_size)
+    data_num = data.shape[0]
+    # split to feature and label
+    x = data.iloc[:, 1:].copy()
+    y = data.iloc[:, 0].copy()
+    return x, y, data_num
+
+
+class FedSKLearnLinearHiggsExecutor(FedSKLearnLinearExecutor):
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/higgs_sklearn_linear_executor.py:33]: Same here, only the customized data loading part, following all our existing example's pattern (CIFAR-10, prostate, xgboost, etc. etc.). We may extend it to handle 
all csv (pandas-handleable) dataset?

[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/higgs_sklearn_linear_executor.py]: applied the same refactor as KMeans

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+            self.local_model.classes_ = np.array([i for i in range(n_classes)])
+            self.local_model.coef_ = np.zeros((1, self.n_features))
+            if fit_intercept:
+                self.local_model.intercept_ = np.zeros((1,))
+        # Training starting from global model
+        # Note that the parameter update using global model has been performed
+        # during global model evaluation
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            self.local_model.fit(x_train, y_train)
```
Comment (YuanTingHsieh) [inline examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:86]: why do we want to catch and ignore warnings here?

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        if self.local_model.fit_intercept:
+            params = {
+                "coef": self.local_model.coef_,
+                "intercept": self.local_model.intercept_,
+            }
+        else:
+            params = {"coef": self.local_model.coef_}
+        return copy.deepcopy(params), self.local_model
```
Comment (YuanTingHsieh) [inline examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:94]: this is returning a tuple of (params, self.local_model)
is this expected?

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    def evaluate(self, curr_round: int, global_param: Optional[dict] = None) -> dict:
+        # set local model with global parameters
+        self.set_parameters(global_param)
+        # perform validation
+        (x_valid, y_valid, valid_size) = self.valid_data
+        y_pred = self.local_model.predict(x_valid)
+        auc = roc_auc_score(y_valid, y_pred)
+        metrics = {"AUC": auc}
+        return metrics, self.local_model
```
Comment (YuanTingHsieh) [inline examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:104]: this is returning a tuple of (metrics, self.local_model)
is this expected?

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/config/config_fed_client.json [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    }
+  ],
+  "task_result_filters": [],
+  "task_data_filters": [],
+  "components": [
+    {
+      "id": "linear_learner",
+      "path": "linear_learner.LinearLearner",
+      "args": {
+        "data_path": "/tmp/nvflare/sklearn/dataset/linear_dataset.csv",
```
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/config/config_fed_client.json:23]: this is a "generic placeholder", which need to be changed according to the specific dataset used, in this case the higgs path

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 2 above)
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:86]: sklearn linear model will print some irrelevant warnings that they haven't removed yet, so I hide them for cleanness

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 3 above)
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:94]: yes the learner expect it

Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:94]: yes the executor expect it https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/app_opt/sklearn/sklearn_executor.py

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 4 above)
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:104]: yes the executor expect it https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/app_opt/sklearn/sklearn_executor.py

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 3 above)
Comment (YuanTingHsieh) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:94]: @ZiyueXu77 got it, then we want to change the type hint for the return in method signature

Should be 

train(xxxxxxx) -> Tuple[dict, dict?]

Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:94]: agree, good catch! will make the change

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 2 above)
Comment (YuanTingHsieh) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:86]: Can you provide what is the warning message?

I am thinking from our side we don't do these catching so when sklearn fix them we would know? 

Also, there might be other warning messages from sklearn that could be useful?

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/config/config_fed_client.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/config/config_fed_client.json:23]: 1. Are we explaining this in the readme? 

2. Do we have scripts that automatically change this to HIGGS for this example?

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 2 above)
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:86]: yes, good point, let me just remove it (my initial purpose is just that I didn't want to see them every time lol)

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py (hunk 4 above)
Comment (YuanTingHsieh) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:104]: Same here for the return type hint

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/config/config_fed_client.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (ZiyueXu77) [reply examples/sklearn-linear/job_configs/sklearn_linear_base/app/config/config_fed_client.json:23]: yes we have it in prepare_job_config, will make it more explicit in readme

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        if self.local_model.fit_intercept:
+            params = {
+                "coef": self.local_model.coef_,
+                "intercept": self.local_model.intercept_,
+            }
+        else:
+            params = {"coef": self.local_model.coef_}
+        return copy.deepcopy(params), self.local_model
+
+    def evaluate(self, curr_round: int, global_param: Optional[dict] = None) -> dict:
```
Comment (YuanTingHsieh) [inline examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:93]: return type hint: `Tuple[dict, dict]`

File: examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        self.n_samples = data["train"][-1]
+        self.n_features = data["train"][0].shape[1]
+        # model will be created after receiving gobal parameters
+
+    def set_parameters(self, params):
+        self.local_model.coef_ = params["coef"]
+        if self.local_model.fit_intercept:
+            self.local_model.intercept_ = params["intercept"]
+
+    def train(self, curr_round: int, global_param: Optional[dict] = None) -> dict:
```
Comment (YuanTingHsieh) [inline examples/sklearn-linear/job_configs/sklearn_linear_base/app/custom/linear_learner.py:59]: return type hint: `Tuple[dict, dict]`

**Review by YuanTingHsieh — APPROVED** [review]
Thanks, LGTM.

**Pull request merged**